[Dart 3] Class modifiers (#4754)
dart-lang/site-www: src/language/class-modifiers.md

---
title: Class modifiers
description: Modifier keywords for class declarations to control external library access.
---

Class modifiers control how a class or a mixin can be implemented, extended,

> or mixed in. 

Reviewer: maybe mention exhaustiveness?  "or mixed in, and whether they are considered for exhaustiveness checking in switches.".

Author: Hmm... Considering your comments, I'm actually leaning towards making this statement even more generic, something like: "Class modifiers control how a class or mixin can be _utilized_ by libraries other than the one they're defined in."

And then letting each modifier's section define what that "utilization" means to it. Or maybe that sounds terrible, or is that technically inaccurate? 

> maybe mention exhaustiveness?

This is a tricky one. My idea is that exhaustiveness checking is technically a product of the fact that `sealed` classes' ability to be utilized by outside classes is restricted, right? So, the intro statement is technically not excluding exhaustiveness checking, even though it is omitting it.

I wanted the purpose of this page to be kept really tight and just define the class modifiers, as agnostic from other topics as possible. Which is tough, considering the purpose of `sealed`. But in the end that decision was made specifically _because_ of `sealed`: because it would be easy for the definition of exhaustiveness checking to spread (and ultimately, be weakened because there'd be no single source of truth) across three different docs (this one, switch, patterns). 

[There will be dedicated content on exhaustiveness checking](https://github.com/dart-lang/site-www/issues/4635) added to the control flow section of the language tour that'll link to and from the `sealed` section here. I'll definitely be incorporating your reviews on emphasizing exhaustiveness checking in the `sealed` section on this page. But, I'm not sure I want to mention anything that specific at the top of this page, muddying what's technically the purpose statement of the page. I might change my mind as I go through the rest of these reviews though.

The full set of modifiers that can appear before a class declaration are `base`,

> `final`, `interface`, `sealed`, `abstract`, and `mixin`. 

Author: I think I took the spec a little too literally by including `mixin` here, but here's my reasoning (which might just be a misunderstanding on my part):

The spec says the new terminology `mixin class` is the _modifier_ `mixin` _on_ `class`, whereas your `mixin M {}` example is a pure mixin declaration. So this page talks about the former, the `mixin` modifier on `class`, while the latter has its own dedicated page for mixin as a standalone construct. 

But even if that is correct(?), I can see that the distinction between `mixin` as a modifier and a pure mixin is too much of an implementation detail on our side and doesn't matter to users, so I'll fix that here.

> `final`, `interface`, `sealed`, `abstract`, and `mixin`. 

Reviewer: FWIW, I think of `mixin` as a modifier when used on a class, and I expect our users will too.

Only the `base` modifier can appear before a mixin declaration. The modifiers do
not apply to other declarations like `enum`, `typedef`, or `extension`.

When deciding whether or not to use class modifiers, consider the following questions:
1. Is [any access](#no-modifier) fine?
2. If not, what do you want to restrict?
    - [Implementation](#base)?
    - [Inheritance](#interface)?
    - [Both](#final)?

>     - [Subtyping](#sealed) by outside libraries at all?

Reviewer: This feels a little off.  The reason for using sealed is not that you want to prevent subtyping by outside libraries (that's what final is for).  The reason is that you want to enable exhaustiveness.

Reviewer 2: Agree.
Rather than focus on the restrictions, I'd focus on intended use-cases and promises (to yourself or others).

So, when deciding whether to use class modifiers, consider the intended uses of the class, and which behavior the class needs to be able to rely on.

If the class is just an interface, and has only abstract members, it should probably be marked `abstract` and `interface`.
Examples are classes like `Sink` or `Future`, which are pure interfaces that others are free to implement.

(The `interface` modifier *can* be used to prevent access to an actual implementation, and force others to implement it from scratch, but there are other ways to achieve that too, so it's will not be a common use-case.)

With `base` you *ensure* that all subclasses inherit the implementation of this class. That allows you to safely call private members on every object of that type, something which allowing `implements` makes unsafe. To ensure that, no subclass is allowed to be implemented, and all subclasses must themselves be at least marked `base` to ensure that.

You should use `base` when you _need_ to rely subclasses inheriting behavior. You shouldn't add `base` just because extending is the *intended* use, and there is no good reason to implement them, because `base` forces subclasses to also be `base`, and subclasses may not want to prevent implementation.

Example of `base` classes are rare.

The `final` prevents subclassing entirely, allowing the author to safely add incremental changes to the API, and to call instance methods knowing that they haven't been overwritten in a third-party subclass.
(The `base` modifier gives most of the same guarantees, except for subclasses overriding instance members, and a `base` class with no public generative constructor could just as well have been `final`. The `final` modifier includes the effect of `base` and therefore all subclasses must again be marked at least `base`.)

The `sealed` modifier is only used to create a known (enumarable) set of subtypes, such that a switch over those types can be statically ensured to be exhaustive. It does not add any new restrictions on the subtypes of those direct subtypes. The sealed class itself cannot be instantiated.

Author: Thanks @lrhn, this is great. I'll be incorporating your points throughout the page, especially the angle of use cases over restrictions.

3. Then, does it need to be [abstract](#abstract)?
4. And lastly, should it be a [class, mixin, or mixin class](#class-mixin-or-mixin-class)?

## No modifier

Use a `class` or `mixin` declaration without a modifier to **allow unrestricted

> access** to outside libraries. The standard behaviors apply:

Reviewer 3: Maybe we should specify "outside its library" more often, or earlier. It doesn't seem to be mentioned until this point which seems a little far in. Maybe something on line 6?

Author: I really like the phrase "permission to create subtype relationships" as a summary for what all the modifiers, in general, are doing. I'll try to use that wherever it makes sense.

> - [Construct][] new instances of a class

Reviewer 4: I found it a little confusing to have these written in the imperative. This is a list of default *capabilities*, not *obligations*. Similarly, I wouldn't say these "behaviors apply". Maybe change that line to something like "By default, you can:" or "This allows you to:"?

Grammar nit: As written, these bullets look like sentences with missing periods.

Author: I like "By default...", thanks for advising!

- [Extend][] a class to create a new subtype

> - [Implement][] a class's interface

Reviewer: a class or mixin's interface.

- [Mix in][] a mixin or mixin class

## `base`

> Use the `base` modifier to **disallow implementation** of a class or mixin.

Reviewer: I don't know if you want to get it into this or not, but it technically only disallows implementation outside of the current library.

> Use the `base` modifier to **disallow implementation** of a class or mixin.

Reviewer 2: I'd focus on the other direction: It *enforces inheritance* of the class or mixin's implementation.

Preventing implementation is a means toward that goal, not a goal in itself.
It's probably mainly a philosophical distinction, but it underlines *why* you'd add `base`. It's not to prevent implementation. If there were other ways to create a subclass which doesn't inherit the superclass implementation, preventing implementation wouldn't give you anything useful by itself.
There isn't, which is why `base` only prevents implementation. That's enough to reach its goal of ensuring that all subclasses inherit the implementation.

> A base class can be extended and constructed. This applies trasitively to all

Reviewer 3: Nit: Typo on `transitively`.
But also, I'm assuming you're talking about disallowing implementation. It feels like this sentence would flow better coming out of the first sentence.

> A base class can be extended and constructed. This applies trasitively to all

Reviewer 2: This also sounds like a `base` class cannot be `abstract`, but that's an orthogonal modifier. An `abstract base class` cannot be constructed.

I'd not discuss construction in the section about `base`, because it's mostly unrelated.

Author: Do you think _all_ the statements about what _isn't_ restricted for each modifier should just be removed? Like this line, or "An interface class _can_ be implemented and constructed.", etc.

subtypes, because implementing a subtype implements the supertype as well.  

{{site.alert.note}}
  Only the `base` modifier can accompany a mixin declaration. 
  - A `base mixin` can be mixed in, but not implemented.
  - A `base mixin class` can be constructed, extended, and mixed in, but not
  implemented.
{{site.alert.end}}

```dart
base class Vehicle {
  String make;
  String model;
  void moveForward(int meters) { … }
}

> var myCar = Vehicle(     // Can be constructed

Reviewer: The class as defined doesn't have a constructor with this signature.  Is this intentional?  It feels better to me if someone could copy this code into an IDE and observe the results themselves.

Author: > Is this intentional?

I just copied these examples from some presentation slides; is there a reason that it'd be intentionally written incorrectly? Maybe to prove some other point? I'm not sure. I wrote this in another reply, but would the correct way be:
```dart
interface class Vehicle {
  String make?;
  String model?;
  void moveForward(int meters) { … }
}
// ...
var myCar = Vehicle();        // Can be constructed
```

Reviewer: To be honest, I don't think there's a lot of value in showing an example of construction.  Construction is controlled orthogonally - it has nothing to do with `interface` or `base` etc.  It's controlled by `abstract`.  So I personally would just leave the constructor part off.  But if you feel that it's important to call out that in the absence of `abstract` an `interface class` can be constructed, then yes, what you have now works.  Or you could just leave off the `make` and `model` fields - they don't add much.

  make: 'Ford',
  model: 'T',
);

class Car extends     // Can be extended
  Vehicle {
    int passengers;
…
}

class MockVehicle     // ERROR: Cannot be implemented

>   implements Vehicle {

Reviewer 2: This class is (apparently) in the same library as the `Vehicle` class, so it *can* implement `Vehicle` (but it must still be marked `base`, `final` or `sealed`, like all subtypes of a `base` declaration must.)

Author: Sorry, should be in a different class. Does that mean that the preceding example (`class Car extends ...`) should actually be (`base class Car extends ...`) to be accurate?

    @override
    void moveForward …
}

```

Disallowing implementation makes `base` useful for:

> - Preventing users of the API from overriding the constructor, for example, to

Reviewer: I don't think overriding is the right term here, constructors aren't virtual.  Something like "Ensuring that the base class constructor is called whenever an instance of a subtype of the class is created"

bypass its validation.

> - Ensuring all private members are available to subtypes of the base class, and

Reviewer: available to -> available on

and accessing them will never cause a [`noSuchMethod`][] error.

## `interface`

Use the `interface` modifier to **disallow extending a class**.
An interface class can be implemented and constructed.

```dart
interface class Vehicle {
  String make;
  String model;
  void moveForward(int meters) { … }
}

> var myCar = Vehicle(     // Can be constructed

Reviewer: As above, no such constructor.

Reviewer 2: And it's needed since `String make;` is not nullable, and therefore needs to be initialized during construction.

Author: What would be the simplest and shortest way to correct this for all the examples? Ideally not distracting from the can/can't be constructed point. Something like this:
```dart
interface class Vehicle {
  String make?;
  String model?;
  void moveForward(int meters) { … }
}
// ...
var myCar = Vehicle();        // Can be constructed

```
?

  make: 'Ford',
  model: 'T',
);

class Car extends     // ERROR: Cannot be inherited
  Vehicle {
    int passengers;
…
}

class MockVehicle     // Can be implemented     
  implements Vehicle {
    @override
    void moveForward …
}
```

Disallowing inheritance makes `interface` useful for:
- Defining API interfaces.

> - Preventing users of the API from overriding a subset of methods.

Reviewer: You could mention the fragile base class problem here (for which interfaces are one solution).

Author: @leafpetersen what is the "fragile base class problem"?

> ## `final` 

Reviewer 3: Regarding your major concerns in the PR:

> Having the code together like that in the same code block makes it seem like it's all in the same library.

I only read this after I commented and I agree. I think it would be useful to not have them in the same code block or to create some representation of different libraries.

> But, if it's all in the same library, would attempting to extend the final class actually be an error? Or would it only be an error if it was attempted in an external library?

It's only an error if attempted in an external library.
You have (almost) full capabilities if using a final/base/interface/sealed class in the same library.

For example, implementing a final class within the same library -- https://github.com/dart-lang/sdk/blob/main/tests/language/class_modifiers/final/final_class_implement_lib.dart
And then once you go outside that library and try to implement a final class -- https://github.com/dart-lang/sdk/blob/main/tests/language/class_modifiers/final/final_class_implement_outside_error_test.dart

> Use the `final` modifier to **disallow extending _and_ implementing** a class.

Reviewer: Might be worth saying that the transitive base property applies here as well?  That is, something like "within the same library, a final class can be extended or implemented, but the subtype is required to be marked base".  Or are you avoiding talking about the "same library" exceptions entirely?

Author: I unintentionally avoided talking about the same library, because I was unclear on whether that functionality was something worth talking about, but now I've got it! I'll add this, thanks

Final classes can be constructed.

```dart
final class Vehicle {
  String make;
  String model;
  void moveForward(int meters) { … }
}

> var myCar = Vehicle(     // Can be constructed

Reviewer: As above.

  make: 'Ford',
  model: 'T',
);

class Car extends     // ERROR: Cannot be inherited
  Vehicle {
    int passengers;
…
}

class MockVehicle     // ERROR: Cannot be implemented     
  implements Vehicle {
    @override
    void moveForward …
}
```

Disallowing inheritance and implementation makes `final` useful for:
- Closing the type hierarchy.
- Ensuring that adding new members to the class won't risk breaking changes for 
its consumers.
- Ensuring access to private members on instances of the final class will never
throw a [`noSuchMethod`][] error.

## `sealed`

> Use the `sealed` modifier to define a class that **cannot be constructed, extended,

Reviewer: This feels like the wrong lead.  I would switch the order of the first two paragraphs. That is, say something like "Use the `sealed` modifier to define a family of classes within a single library for which switches will be checked for exhaustiveness.  In order to allow the compiler to check exhaustiveness, applying `sealed` to a class will prevent it from being extended or implemented outside of the library where the class is defined."

> or implementated** outside of the library where the class is defined. 

Reviewer 3: Nit: Typo with `implemented`

> The purpose of the `sealed` modifier is to enable [_exhaustiveness checking_][]

Reviewer 3: I would maybe start with this information -- that the main point of `sealed` is for exhaustiveness. The fact that it can't be subtyped at all outside of its library comes after the exhaustiveness imo.

We prefer users to use `final` if they don't need the exhaustiveness feature.

for switch cases.
Any possible direct subtypes of a sealed type can only exist in the same library,

> so the compiler is always aware of the extent of a sealed type's hierarchy.

Author: Thank you for explaining that!

  String make;
  String model;
  void moveForward(int meters) { … }
}

class Car extends Vehicle { }
class Truck implements Vehicle { }
class Bicycle extends Vehicle { }

// ...

return switch (vehicle) {     // ERROR: The switch is missing a subtype of Vehicle
  Car => 'vroom',
  Truck => 'VROOOOMM'
};
```

## `abstract`

Use the `abstract` modifier to define a class that can’t be constructed.
Abstract classes are useful for defining interfaces, often with some
implementation. If you want your abstract class to appear to be instantiable,
define a [factory constructor][].

Abstract classes often have [abstract methods][].
Here’s an example of declaring an abstract class that has an abstract
method:

<?code-excerpt "misc/lib/language_tour/classes/misc.dart (abstract)"?>
```dart
// This class is declared abstract and thus
// can't be instantiated.
abstract class AbstractContainer {
  // Define constructors, fields, methods...

  void updateChildren(); // Abstract method.
}
```

## `class`, `mixin`, or `mixin class`?

The `mixin` declaration defines a [mixin][]. The `mixin` _modifier_ can only be
applied to a `class` declaration to create a _mixin class_: a class that can be

> treated both as a class and a mixed in. Regular classes cannot be mixed in.

Reviewer 3: Nit: Typo on `treated both as a class and a mixed in`.

Author: I think I will end up doing that, but I'm also going to add some mixin examples to this page.

> The `mixin` declaration defines a [mixin][]. The `mixin` _modifier_ can only be
> applied to a `class` declaration to create a _mixin class_: a class that can be
> treated both as a class and a mixed in. Regular classes cannot be mixed in.

Reviewer 5: Again, I don't think it's helpful to introduce the notion of "a class that can be mixed in". The mixin which is introduced by `... mixin class MC ...` is a perfectly normal mixin, and the class is a perfectly normal class, and we don't need to give rise to questions about "how mixin application works when we apply a class that can be mixed in" etc.etc.

Perhaps:
```suggestion
A `mixin` declaration defines a [mixin][]. A `class` declaration defines a [class][]. A
`mixin class` declaration defines both a mixin and a class, with the same name and
the same type.
```

Author: That's perfect, thank you for simplifying it

Reviewer: Hmm, I don't agree with the suggested rewrite here.  I don't know what it means to have a class and a mixin in scope with the same name and the same type.  I think the original phrasing is both more accurate and less confusing.

> 

Reviewer 3: But also, maybe an example of a mixin class would be useful? Like how we do with the other modifiers.

Mixins and mixin classes are meant to be mixed in, so they cannot be combined

> with the modifiers that prevent inheritance: `interface`, `final`, and `sealed`.

Author: Thanks for this, especially the definition of what mixins are intended for. Unrelated to the rest of this PR, I felt that statement was missing from the original mixins page, so I added it there.

## Combining modifiers

It is possible to combine some modifiers for layered restrictions. For example, 
an `abstract interface class` combines restrictions on instantiation and inheritance.
This could be useful to define a "pure contract", or a class that doesn't update
any preexisting state.

Modifiers that are mutually exclusive cannot be combined. For example,
`base` allows extending and disallows implementing, and `interface` allows
implementing and disallows extending.

Modifiers that are redundant cannot be combined. For example, mixins and sealed
classes already cannot be constructed, so combining them with `abstract`, which

> disallows construction, is redundant. 

Reviewer 2: This way of describing it, as if all modifiers are the same and the rules for how they can be combined are complicated, gets me very confused about something that's fairly simple (to me).

A class declaration can be, in order:
* Optionally an `abstract`, describing whether the class can contain abstract members (and it also prevents instantiation).
* Optionally one of `base`, `interface`, `final` or `sealed`, describing restrictions on other libraries subtyping the class.
* Optionally a `mixin`, describing whether the declaration also introduces a mixin (aka. can be mixed in).
* and then the `class` keyword itself.

Some combinations are disallowed:
* `abstract` with `sealed` (because a `sealed` class is always implicitly `abstract`).
* `interface`, `final` or `sealed` with `mixin` (because those access modifiers prevent mixing in).

The phrasing above doesn't say that `abstract` has to come first, and it sounds like combining `abstract` and `base` is at the same level as combining `base` and `interface`. To me it isn't. There are different groups of modifiers that can be combined (almost) orthogonally.

Author: That's a lot clearer, I think I'll rewrite the entire section to more or less match this comment

> The [syntax specification] documents the valid combinations and their capabilities.

Reviewer 2: As with all writing, it depends on who the intended audience is.
I'd consider these documents as introductory. They are not exhaustive in defining or describing the language, so I would focus in the common use-cases, for the syntax, to *show* how it's used, not *specify* it.

(We have a secondary problem that we don't have a good, complete and readable exhaustive description of the language. The specification fails, at least, the "readable" part, because it mixes intricate semantic details with the "what does this language construct really do" text, and is very short on examples.)

> The [syntax specification] documents the valid combinations and their capabilities.

Author: > We have a secondary problem that we don't have a good, complete and readable exhaustive description of the language

Yeah that issue is definitely on my radar. I'd like to start developing a "reference" type of documentation, especially now with patterns and class modifiers. Hopefully after this release 🤞

> The [syntax specification] documents the valid combinations and their capabilities.

Author: I made an issue to track the need for further reference for this topic: https://github.com/dart-lang/site-www/issues/4783

---
title: Class modifiers
description: Modifier keywords for class declarations to control external library access.
---

> Class modifiers control how a class or a mixin can be subtyped in libraries

Reviewer 6: except for `abstract`, right?

Reviewer 6: I'm not sure... maybe the most clear explanation is to explicitly call out that `abstract` applies to all classes regardless of the library, and the rest of the modifiers control how a class or mixin can be subtyped.

Reviewer 7: The `mixin` modifier applies to code within the same library too.

> outside of the library where it's defined.

Reviewer 6: Ah, thanks. It would be great to add this definition to the docs if we haven't already.

> outside of the library where it's defined.

Author: We have a new Glossary page that this definition seems suited for! Thanks, I'll add that

> outside of the library where it's defined.

Author: Created https://github.com/dart-lang/site-www/issues/4784 to make sure this gets added and linked from here.

The full set of modifiers that can appear before a class declaration
are `abstract`, [`mixin`][mixin], `base`, `final`, `interface`, and `sealed`. 
Only the `base` modifier can appear before a mixin declaration. The modifiers do
not apply to other declarations like `enum`, `typedef`, or `extension`.

When deciding whether to use class modifiers, consider the intended uses of the
class, and what behaviors the class needs to be able to rely on.

## No modifier

Use a `class` or `mixin` declaration without a modifier to allow unrestricted

> permission to create subtype relationships from outside libraries.

Reviewer 6: To me, it's a little odd to say "from outside libraries" here, since `abstract` doesn't place restriction based on the library, it applies to all code that tries to subclass it.

```suggestion
permission to create subtype relationships from any library.
```

Author: Yeah... maybe moving that here was a mistake because it makes it harder to summarize the whole page. I'll fix it, but need to make sure that "outside libraries" is explicitly stated for all of the ones that it does affect

## `abstract`

Use the `abstract` modifier to define a class that can’t be constructed.
Abstract classes are useful for defining interfaces, often with some
implementation. Abstract classes often have [abstract methods][]:

```dart
// Library a.dart
abstract class Vehicle {
  
  void moveForward();     // Abstract method.
}
```

```dart
// Library b.dart
import 'a.dart';

var myCar = Vehicle();     // Error: Cannot be constructed

class Car extends          // Can be extended
  Vehicle {
    int passengers;
    ...
}

class MockVehicle          // Can be implemented
  implements Vehicle {
    @override
    void moveForward(int meters) { ... }
}
```

If you want your abstract class to appear to be instantiable,
define a [factory constructor][].

## `base`

Use the `base` modifier to enforce inheritance of a class or mixin's implementation.

The `base` modifier disallows implementation outside of the current library, which
guarantees:

- The base class constructor is called whenever an instance of a subtype of the
class is created.
- That all implemented private members exist in subtypes.
- Adding new members to a class reduces the risk of breaking subtypes, since all

> subtypes will inherit the new member.

Author: Ah, I see. Thanks for catching that, very subtle!

base class Car extends          // Can be extended
  Vehicle {
    int passengers;
    ...
}

base class MockVehicle          // ERROR: Cannot be implemented
  implements Vehicle {
    @override
    void moveForward ...
}
```

## `interface`

Use the `interface` modifier to define an interface. Interface classes can be
implemented outside of the current library, but not extended. This guarantees:

- Users of the API are prevented from overriding a subset of methods.
- A solution to the [fragile base class problem][], or the freedom to change the
interface's implementation, because there are no inherited member implementations
outside of the current library whose behavior needs to be preserved. 

```dart
// Library a.dart
interface class Vehicle {
  void moveForward(int meters) { ... }
}
```

```dart
// Library b.dart
import 'a.dart';

var myCar = Vehicle();     // Can be constructed

class Car extends          // ERROR: Cannot be inherited
  Vehicle {
    int passengers;
    ...
}

class MockVehicle          // Can be implemented     
  implements Vehicle {
    @override
    void moveForward ...
}
```

To define a pure interface that others are free to implement, that cannot be

> inherited and only has abstract members, define an `abstract interface class`.

Author: I added the line "All members of the class must have an implementation." to the list of guarantees at the beginning of the `interface` section, if that makes sense?

## `final` 

Use the `final` modifier to prevent subtyping from a class outside of the current
library, closing the type hierarchy. 

Disallowing both inheritance and implementation means `final` classes provide all
the same guarantees of [`base`](#base) and [`interface`](#interface) classes.

Final class can be extended or implemented within the

> same library, but like the `base` modifier, any subtypes must be marked `base`.

Reviewer 5: Perhaps:
```suggestion
same library, but like the `base` modifier, any subtypes must be marked `base`,
`final`, or `sealed`.
```
The point would be that it is misleading to tell the reader that subtypes of a final class must be `base`. For instance, it makes sense that they would declare every subtype of a given final class `final` as well, thus creating an entirely `final` subtype hierarchy. That hierarchy is completely closed, in the sense that there will never be any subtypes that we don't know about (for instance, if the classes aren't generic then we can use `o.runtimeType == something` to deal with each of those types precisely, which is even stronger than `sealed`). So we shouldn't imply that this is impossible.

---
title: Class modifiers
description: Modifier keywords for class declarations to control external library access.
---

> Class modifiers control how a class or a mixin can be subtyped, both

Author: This is a tough one because it's typically outside libraries (so, not "you", probably someone else) attempting the subtyping. So you'd have to say "how _other libraries_ can subtype...". BUT, it's not exclusively about _other_ libraries, because it could also be from within the same library... Maybe:

> Class modifiers control _how a class or a mixin allows subtyping,_ both
[from within its own library](#abstract), and from outside of the library where
it's defined.

At this point I'm scared to change anything that hasn't undergone technical review, but I think this is safe... 😬

[from within its own library](#abstract), and from outside of the library where
it's defined.

> Modifiers prepend a class or mixin declaration.

Author: I don't want to make it passive, but I'm also hesitant to say "_You_ can prepend class modifiers...". I'm not sure exactly what you mean by the subject and object being backwards so I can't think of another way to say it besides "Modifiers _come before_ a class or mixin declaration"?

> Modifiers prepend a class or mixin declaration.

Reviewer 7: > I'm not sure exactly what you mean by the subject and object being backwards

As far as I know, to prepend means to put something before another thing. So "Modifiers prepend a class or mixin declaration" means that a modifier goes out and puts something (unspecified) before a class or mixin declaration.

> "Modifiers come before a class or mixin declaration"?

👍

> Modifiers prepend a class or mixin declaration.

Author: I'll fix this in the PR for #4826

> Modifiers prepend a class or mixin declaration.

Reviewer 2: (As a third-language participant, I understand "A prepends B with C" to mean Subject A takes Object B and puts Object C before it. If we skip any of the parts, it becomes unclear which one is missing. "C prepends B" can be read as any of:
* Object C is put before Object B by unknown subject. (Matches non-transitive verbs like "predates".)
* Subject C puts Object B before unknown object.
* Subject C puts unknown object before Object B.

For example, writing `abstract class` defines an abstract class.
The full set of modifiers that can appear before a class declaration are:

- `abstract`
- `base`
- `final`
- `interface`
- `sealed`
- [`mixin`][class, mixin, or mixin class]

Only the `base` modifier can appear before a mixin declaration. The modifiers do
not apply to other declarations like `enum`, `typedef`, or `extension`.

When deciding whether to use class modifiers, consider the intended uses of the
class, and what behaviors the class needs to be able to rely on.

## No modifier

Use a `class` or `mixin` declaration without a modifier to allow unrestricted
permission to create subtype relationships from any library.
By default, you can:

- [Construct][] new instances of a class.
- [Extend][] a class to create a new subtype.
- [Implement][] a class or mixin's interface.
- [Mix in][mixin] a mixin or mixin class.

## `abstract`

Use the `abstract` modifier to define a class that can’t be constructed from any

> library (it's own, as well as outside libraries).

Reviewer 8: i would say the reason you use abstract isn't to prevent construction so much as to allow you to not provide a full concrete implementation of the entire interface.

Author: Changed to:
> Use the `abstract` modifier to define a class that doesn't require a full, concrete
implementation of the entire interface.

Abstract classes are useful for defining interfaces, often with some
implementation. Abstract classes often have [abstract methods][].

```dart
// Library a.dart
abstract class Vehicle { 
  void moveForward();     // Abstract method.
}
```

```dart
// Library b.dart
import 'a.dart';

var myCar = Vehicle();     // Error: Cannot be constructed

class Car extends          // Can be extended
  Vehicle {
    int passengers;
    ...
}

class MockVehicle          // Can be implemented
  implements Vehicle {
    @override
    void moveForward(int meters) { ... }
}
```

If you want your abstract class to appear to be instantiable,
define a [factory constructor][].

## `base`

Use the `base` modifier to enforce inheritance of a class or mixin's implementation.
The `base` modifier disallows implementation outside of the current library, which
guarantees:

- The base class constructor is called whenever an instance of a subtype of the
class is created.
- That all implemented private members exist in subtypes.
- Adding a new implemented member to a `base` class will not break subtypes (except
  when they already declare a member with the same name and an incompatible

>   signature), since all subtypes will inherit the new member.

Reviewer 8: that parenthetical is doing a lot of work!

Author: Made it a sub bullt

>   signature), since all subtypes will inherit the new member.

Author: Can't really active-voice the first bullet again since it could be anyone, most likely not "you", when we're talking about restricting external libraries. Other notes incorporated though

The `base` modifier must be applied transitively to all the subtypes of a class;
any class which implements or extends a base class must be marked `base`, `final`,

> or `sealed`. 

Reviewer 8: we should say why

> or `sealed`. 

Author: Shortened the sentence, not sure what the condition/instruction is here

```dart
// Library a.dart
base class Vehicle {
  void moveForward(int meters) { ... }
}
```

```dart
// Library b.dart
import 'a.dart';

var myCar = Vehicle();          // Can be constructed

base class Car extends          // Can be extended
  Vehicle {
    int passengers;
    ...
}

base class MockVehicle          // ERROR: Cannot be implemented
  implements Vehicle {
    @override
    void moveForward ...
}
```

## `interface`

Use the `interface` modifier to define an interface. Interface classes can be
implemented outside of the current library, but not extended. This guarantees:

> - All members of the class must have an implementation.

Reviewer 8: that's actually quite surprising, if you're coming from other languages with an `interface` keyword. (Most other languages' `interface` keyword means what we mean by `abstract interface class`.) Might be worth acknowledging this weirdness.

Reviewer 5: There are some sources of ambiguity here. 'All members of the class' would generate the question 'Which class?'.

You can certainly have an `abstract interface class A ...` where some members do not have an implementation. With a plain `interface class C` it is required that `C` is a concrete class (simply because we didn't say `abstract`), and in that case all its members must have an implementation.

But this is all about `abstract` or not, which means that it is confusing to mention it as if it had something to do with `interface`.

I guess the point is that we're allowed to have implementations of members in an `interface class` and we're still guaranteed (assuming no reopening) that nobody can inherit those member implementations. This means that we have more freedom to change that implementation.

In other languages you'd probably write an `interface I` and then `private class C implements I`, and that's how you would ensure that nobody can inherit the instance member implementations. With a Dart `interface class` we can do both in the same declaration. This may be seen as ugly or not, but you will at least be able to avoid duplicating a bunch of member signatures (that is: the parameter types, return type and name, for each member).

> - Users of the API are prevented from overriding a subset of methods.

Reviewer 8: not sure what this line means

- A solution to the [fragile base class problem][], or the freedom to change the
interface's implementation, because there are no inherited member implementations

> outside of the current library whose behavior needs to be preserved. 

Reviewer 8: this doesn't sound like a guarantee, it sounds like a reason to use this setup.

```dart
// Library a.dart
interface class Vehicle {
  void moveForward(int meters) { ... }
}
```

```dart
// Library b.dart
import 'a.dart';

var myCar = Vehicle();     // Can be constructed

class Car extends          // ERROR: Cannot be inherited
  Vehicle {
    int passengers;
    ...
}

class MockVehicle          // Can be implemented     
  implements Vehicle {
    @override
    void moveForward ...
}
```

To define a pure interface that others are free to implement, that cannot be
inherited but is allowed to have abstract members, define an

> `abstract interface class`.

Reviewer 8: i feel like we should emphasize this combination more. maybe even having this as a top-level section before the existing `interface class`?

## `final` 

Use the `final` modifier to prevent subtyping from a class outside of the current
library, closing the type hierarchy. 
Disallowing both inheritance and implementation means `final` classes provide all

> the same guarantees of [`base`](#base) and [`interface`](#interface) classes.

Reviewer 8: "the same guarantees of [`base`](#base) and [`interface`](#interface) classes" is a weird framing. I mean, it's true, but I feel it's misleading. When I think of the "guarantees of base" I think "allows me to extend it" and when I think of the "guarantees of interface" I think "allows me to implement it". But "final" means neither of those.

Author: Great point, some of the guarantees cancel each other out. I'll reword this

Author: Changed to:

> Disallowing both inheritance and implementation prevents subtyping entirely.
> This guarantees:
> 
> - You can safely add incremental changes to the API.
> - You can call instance methods knowing that they haven't been overwritten in a
> third-party subclass.

Final classes can be extended or implemented within the
same library, but like the `base` modifier, any subtypes must be marked `base`, 

> `final`, or `sealed`.

Reviewer 8: we should say why

  Vehicle {
    int passengers;
    ...
}

class MockVehicle          // ERROR: Cannot be implemented     
  implements Vehicle {
    @override
    void moveForward ...
}
```

## `sealed`

Use the `sealed` modifier to create a known, enumerable set of subtypes, so that
a switch over those types can be statically ensured to be [_exhaustive_][exhaustive].

The `sealed` modifier prevents a class from being extended or implemented outside
of the library where the class is defined. The compiler is aware of any possible
direct subtypes because they can only exist in the same library.

This allows the compiler to alert you when a switch is not exhaustively handling
all possible subtypes in its cases:

```dart
sealed class Vehicle { ... }

class Car extends Vehicle { }
class Truck implements Vehicle { }
class Bicycle extends Vehicle { }

// ...

var vehicle = Vehicle();                 // ERROR: Cannot be instantiated

// ...

return switch (Vehicle vehicle) {       // ERROR: The switch is missing a subtype of Vehicle
  Car => 'vroom',
  Truck => 'VROOOOMM'
};
```

If [exhaustiveness checking][exhaustive] is not necessary for your use case, use

> [`final`](#final) instead.

Reviewer 8: we should talk about how `sealed` implies `abstract` and why that is

## Combining modifiers

> It is possible to combine some modifiers for layered restrictions. 

Author: Didn't know second person is preferred, that'll come in handy

> A class declaration can be, in order:

Reviewer 9: **question:** If this is in order, why is the list not an ordered list?

* (Optional) `abstract`, describing whether the class can contain abstract members
and prevents instantiation.
* (Optional) One of `base`, `interface`, `final` or `sealed`, describing
restrictions on other libraries subtyping the class.
* (Optional) `mixin`, describing whether the declaration can be mixed in.
* The `class` keyword itself.

> Some combinations are disallowed beacause they are contradictory, redundant, or
> otherwise mutually exclusive:

Reviewer 9: **issues:** Write in active voice per [GSG](https://developers.google.com/style/voice). 
Write as shorter sentences per [GSG](https://developers.google.com/style/accessibility#ease-of-reading). (Reading: 19.4 > 7.1)
**typo:** `because`, not `beacause`.
```suggestion
You can't combine some modifiers. They include the following:
```

> * `abstract` with `sealed` (because a sealed class is always implicitly abstract).
> * `interface`, `final` or `sealed` with `mixin` (because those access modifiers
> prevent mixing in).

Reviewer 9: **issue:** Avoid parentheses per [GSG](https://developers.google.com/style/parentheses). 
Write in active voice per [GSG](https://developers.google.com/style/voice).

```suggestion
* `abstract` with `sealed`. A sealed class is always abstract.
* `interface`, `final` or `sealed` with `mixin`.
    These access modifiers prevent mixing in.
```

**suggestion:** Link `sealed class` to the `sealed` anchor earlier in the page. Also, link to `mixin` if the modifiers are mentioned on that page.

[class, mixin, or mixin class]: /language/mixins#class-mixin-or-mixin-class
[mixin]: /language/mixins
[fragile base class problem]: https://en.wikipedia.org/wiki/Fragile_base_class
[`noSuchMethod`]: /language/extend#nosuchmethod
[construct]: /language/constructors
[extend]: /language/extend
[implement]: /language/classes#implicit-interfaces
[factory constructor]: /language/constructors#factory-constructors

> [exhaustive]: /

Reviewer 8: TODO?

Reviewer 8: (if it was intended as a joke, it's a funny joke, but maybe one too subtle for people to get when what they're looking for is information on exhaustiveness checking!)

## `base`

To enforce inheritance of a class or mixin's implementation, use the `base` modifier.
A base class disallows implementation outside of its own library. This guarantees:

- The base class constructor is called whenever an instance of a subtype of the
class is created.
- All implemented private members exist in subtypes.
- A new implemented member in a `base` class does not break subtypes,
since all subtypes inherit the new member.
  - This is true unless the subtype already declares a member with the same name
  and an incompatible signature.

> Any class which implements or extends a base class must transitively be marked
> `base`, `final`, or `sealed`. This is necessary to prevent outside libraries from
> breaking the base class guarantees.

Reviewer 9: **issue:** This is a little hard to grok.
```suggestion
Mark all classes derived from a base class as `base`, `final`, or `sealed`.
This prevents outside libraries from breaking the base class guarantees.
```

## `interface`

To define an interface, use the `interface` modifier. Libraries outside of the
interface's own defining library can implement the interface, but not extend it.
This guarantees:

> - When an instance method in the class calls another instance method on `this`,
> it will always invoke a known implementation of the method from the same library.
> - Other libraries can't override methods which may then be called by the interface
> class's own methods in unexpected ways. This reduces the [fragile base class problem][].

Reviewer 9: **issue:** Write as shorter sentences per [GSG](https://developers.google.com/style/accessibility#ease-of-reading).
```suggestion
- When a class instance method calls another instance method on `this`,
the first method invokes a known implementation of the method from the same library.
- Other libraries can't override methods that the interface class's own
methods might call later in unexpected ways. This reduces the [fragile base class problem][].
```

}
```

### `abstract interface`

> The most common use for the `interface` modifier is to [combine](#combining-modifiers)
> it with [`abstract`](#abstract) for an `abstract interface class`.
> This defines a pure interface. 

Reviewer 9: **suggestion:** Write in active voice per [GSG](https://developers.google.com/style/voice).
```suggestion
To create a pure interface, [combine](#combining-modifiers) the `interface` 
and [`abstract`](#abstract)  modifiers. 
The resulting `abstract interface class` defines a pure interface. 
```

Like an `interface` class, other libraries can implement, but cannot inherit,

> a pure interface. Like an `abstract` class, a pure interface is allowed to have

Reviewer 9: **nit:** Simplify the word choice.
```suggestion
a pure interface. Like an `abstract` class, a pure interface can have
```

## `sealed`

> To create a known, enumerable set of subtypes, so that a switch over those types
> can be statically ensured to be [_exhaustive_][exhaustive], use the `sealed` modifier.

Reviewer 9: **issue:** Write as shorter sentences per [GSG](https://developers.google.com/style/accessibility#ease-of-reading).
```suggestion
To create a known, enumerable set of subtypes, use the `sealed` modifier.
This ensures [_exhaustive_][exhaustive] coverage of all `cases` included in a `switch` over those types.
```

The `sealed` modifier prevents a class from being extended or implemented outside

> of the library where the class is defined. They also prevent construction,
> and are therefore implicitly [abstract](#abstract).

Reviewer 9: **suggestion:** Write in active voice per [GSG](https://developers.google.com/style/voice). Not certain of my construction here, so you can ignore.
```suggestion
of the library where you define the class. You can't construct these classes,
so they have implicit [abstraction](#abstract).
```

> The compiler is aware of any possible direct subtypes because they can only exist
> in the same library. This allows the compiler to alert you when a switch does not
> exhaustively handle all possible subtypes in its cases:

Reviewer 9: **suggestion:** Not sure if this is better. I'm not sure if another construction could handle "exhaustively" better. Ignore if desired.
```suggestion
As possible direct subtypes can only exist in the same library,
their limited scope makes the compiler aware of them.
This limitation allows the compiler to alert you when a `switch` does not
exhaust all possible subtypes in its `case` statements:
```

};
```

> If you don't intend users to be able to [exhaustively switch][exhaustive] over
> the subtypes of the class, or you want the freedom to add new subtypes later
> without it being a breaking API change, use [`final`](#final) instead.

Reviewer 9: **issue:** Write as shorter sentences per [GSG](https://developers.google.com/style/accessibility#ease-of-reading).
```suggestion
If you don’t want [exhaustive switching]][exhaustive] or want to add subtypes later,
use [`final`](#final).
```
